### ISSUE TITLE
Immutable (vs mutable) types defined

### ISSUE THREAD

**Issue by guardrex** [dotnet/docs#29187]
I'm seeking a C# Guide definition for "immutable type" for a Blazor article, where I would like to link the text "immutable" to a formal definition. However, I only see "immutable (type)" used to describe types in passing and in other general passing remarks, not a formal definition of "immutable type" versus "mutable type." When "immutable (type)" is used, it's never linked to a definition, including to any non-MS page. Is this by-design, or is it missing coverage that should be added (somewhere, not necessarily to this particular topic)? If by-design, can you recommend a link destination for "immutable (type)" for the Blazor article?

The context for the cross-link is the word "immutable" in the following bullet point ...

> * All of the parameters are of the following immutable types, whether nullable or nonnullable, and haven't changed since the previous set of parameters were set:

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 1991c89a-9728-ea31-a999-79b3cb331372
* Version Independent ID: 9aa4f01b-a5a0-3cb1-3c22-db333c0d9f11
* Content: [The C# type system](https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/types/)
* Content Source: [docs/csharp/fundamentals/types/index.md](https://github.com/dotnet/docs/blob/main/docs/csharp/fundamentals/types/index.md)
* Product: **dotnet-csharp**
* Technology: **csharp-fundamentals**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

**Comment by guardrex** [comment]
I also need to figure out what to do with the word "known," another cryptic word here for newbies. "Known" by who? What is "known" about the type?

Since I also don't see a clean C# definition for "known type," I wonder if I can remark that a known type implements `System.Type`.

I don't think the sentence works well with many embedded parenthetical cross-links. I'm inclined to handle this with ye 'olde dagger ...

> All of the parameters are known immutable types&dagger; or any \[primitive type](/dotnet/api/system.type.isprimitive), whether nullable or nonnullable, and haven't changed since the previous set of parameters were set.
>
> &dagger;An \*immutable type\* is a \[\`readonly\`](/dotnet/csharp/language-reference/keywords/readonly) type that prevents a member from changing any property or field value after the type is constructed. A *known type* is a type that implements \<xref:System.Type?displayProperty=fullName>.

Is that &dagger; remark accurate?

**Comment by BillWagner** [comment]
> Are those † remarks accurate?

No. I should update that language.  I meant that the type is one the compiler can verify as immutable. Those are a few types:

- The builtin unmanaged types (`int`, `char`, `float`, and so on). 
- `readonly struct` types.
- Types where every member field is a type the compiler can verify as immutable and is marked as `readonly`.

That third rule is the one that gets involved. For examples:

```csharp
// This is verifiably immutable: All members are `readonly` and are themselves verifiably readonly:
public struct Point
{
    public readonly double X;
    public readonly double Y;
    public Point(double x, double y) => (X, Y) = (x, y);
}

// This is not verifiably immutable. Its members are readonly, but they are not verifiably immutable
public class Stuff
{
    // OK, the variable `messages` can't change, but the list itself isn't immutable (its contents could change.)
    private readonly List<string> messages = new List<string>();
}
```

**Comment by guardrex** [comment]
Excellent ... thx. Then, you'd feel good if I went with ...

> All of the parameters are \[verifiably immutable types](/dotnet/csharp/language-reference/builtin-types/unmanaged-types) or any \[primitive type](/dotnet/api/system.type.isprimitive) that hasn't changed since the previous set of parameters were set.

**Comment by BillWagner** [comment]
That's good by me @guardrex

**Comment by szalapski** [comment]
I'm confused: What does immutability have to do with unmanaged? References of the unmanaged types are all mutable.

**Comment by BillWagner** [comment]
I read through dotnet/AspNetCore.Docs#25776 and I'm not sure the best resolution.

It might make the most sense to not link between the C# docs and the Blazor docs for this concept. I had been answering from the perspective of the C# language rules. In the language, `readonly struct` types are the only types guaranteed to be immutable. (To your side note, the runtime team has added the `readonly` modifier on may of the unmanaged types (Int32, etc).

If I understand correctly, and it's a set of *known types* for the Blazor ComponentBase, it's probably simplest just to list them. That way you avoid any confusion. 

Would that fix the issue?

**Comment by guardrex** [comment]
Perhaps .....

> \* All of the parameters are known&dagger; \[verifiably immutable types](/dotnet/csharp/language-reference/builtin-types/unmanaged-types) or any \[primitive type](/dotnet/api/system.type.isprimitive) that hasn't changed since the previous set of parameters were set.
>
>  &dagger;Blazor's framework uses a set of built-in rules and explicit parameter types for change detection. These rules and the types are subject to change at any time. For more information, see the \[\`ChangeDetection\` API in the ASP.NET Core reference source](https://github.com/dotnet/aspnetcore/blob/main/src/Components/Components/src/ChangeDetection.cs).

**Comment by BillWagner** [comment]
I'll close this.  I think we've reached the right resolution on the Blazor docs, and the C# language allows, but does not have general syntax to enforce immutable types . (`readonly struct` does enforce immutable struct types, but that's not general to other types).
